File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/constant/number/OctalPart.java
package xyz.wagyourtail.commons.parsers.impl.constant.number;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

public class OctalPart extends StringData.OnlyRaw<Data.SingleContent<String>> {

    public OctalPart(String rawContent) {
        super(rawContent, OctalPart::getContentChecked);
    }

    public OctalPart(CharReader<?> reader) {
        super(reader, OctalPart::getContentChecked);
    }

    public static OctalPart parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new OctalPart(reader);
        reader.expectEOS();
        return number;
    }

    public static OctalPart unchecked(String rawContent) {
        return new OctalPart(rawContent);
    }

    private static SingleContent<String> getContentChecked(CharReader<?> reader) {
        val first = reader.peek();
        if (first < '0' || first > '7') {
            throw reader.createException("Not an octal number: " + (char) first);
        }

        return new SingleContent<>(reader.takeWhile(e -> e >= '0' && e <= '7'));
    }
}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/constant/number/OctalPart.kt
package xyz.wagyourtail.commonskt.parsers.impl.constant.number

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader

class OctalPart(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<OctalPart, SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): OctalPart {
            return OctalPart(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            val first = reader.peek()

            if (first != null && (first >= '0' && first <= '7')) {
                return SingleContent(reader.takeWhile { it >= '0' && it <= '7' })
            }

            throw reader.createException("Not a octal number: $first")
        }

    }

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }

}